lib/netutil/proxyprotocol.go

```diff
@@ -0,0 +1,176 @@
+package netutil
+
+import (
+	"bufio"
+	"bytes"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"sync"
+)
+
+var bufioReaderPool = sync.Pool{}
+
+func getBufioReader(r io.Reader) *bufio.Reader {
+	v := bufioReaderPool.Get()
+	if v == nil {
+		return bufio.NewReader(r)
+	}
+	br := v.(*bufio.Reader)
+	br.Reset(r)
+	return br
+}
+
+func putBufioReader(r *bufio.Reader) {
+	bufioReaderPool.Put(r)
+}
+
+type proxyProtocolConn struct {
+	net.Conn
+	br         *bufio.Reader
+	remoteAddr net.Addr
+}
+
+func newProxyProtocolConn(srcConn net.Conn) (net.Conn, error) {
+	br := getBufioReader(srcConn)
+	maybeSrcAddr, err := readProxyProto(br)
+	if err != nil {
+		return nil, fmt.Errorf("cannot read proxy protocol from connection: %w", err)
+	}
+	if maybeSrcAddr == nil {
+		maybeSrcAddr = srcConn.RemoteAddr()
+	}
+	return &proxyProtocolConn{
+		Conn:       srcConn,
+		br:         br,
+		remoteAddr: maybeSrcAddr,
+	}, nil
+}
+
+func (ppc *proxyProtocolConn) Read(b []byte) (n int, err error) {
+	return ppc.br.Read(b)
+}
+
+func (ppc *proxyProtocolConn) RemoteAddr() net.Addr {
+	return ppc.remoteAddr
+}
+
+func (ppc *proxyProtocolConn) Close() error {
+	putBufioReader(ppc.br)
+	return ppc.Conn.Close()
+}
+
+var (
+	V2Identifier = []byte("\r\n\r\n\x00\r\nQUIT\n")
+)
+
+// https://www.haproxy.org/download/1.8/doc/proxy-protocol.txt
+func readProxyProto(r *bufio.Reader) (net.Addr, error) {
+	// first 12 bytes - signature
+	// 13 byte - protocol and command
+	maybeProto, err := r.Peek(13)
+	if err != nil {
+		if errors.Is(err, io.EOF) {
+			return nil, nil
+		}
+		return nil, fmt.Errorf("cannot read data from buffer: %w", err)
```

Comment:
Can it be less than 13 bytes for non-proxy protocol? If yes, it can respond with an error to a caller...

Reply:
Maybe we shouldn't ever return an error here. Just log that there was an attempt to parse the protocol and it failed.

Reply:
See the comment at https://github.com/VictoriaMetrics/VictoriaMetrics/pull/3687#discussion_r1088658161